Accenture/reactive-interaction-gateway
# HTTP/2!

# What's included

- Upgrade Cowboy
- Drop Phoenix transports
- Add HTTP/2 + HTTPS support
- SSE/WS subscriptions parameter that allows setting up subscriptions along with establishing a connection

Note that the self-signed certificates are not used in production.

Closes #34, #115.

Follow-up: #102 and #128

# How to test this

I've verified HTTP/2 using `curl`:

```plain
curl -kv https://localhost:4001/_rig/v1/connection/sse
*   Trying ::1...
* TCP_NODELAY set
* Connection failed
* connect to ::1 port 4001 failed: Connection refused
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 4001 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* Cipher selection: ALL:!EXPORT:!EXPORT40:!EXPORT56:!aNULL:!LOW:!RC4:@STRENGTH
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/cert.pem
  CApath: none
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Client hello (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS change cipher, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES256-GCM-SHA384
* ALPN, server accepted to use h2
* Server certificate:
*  subject: O=Phoenix Framework; CN=Self-signed test certificate
*  start date: Dec 13 00:00:00 2018 GMT
*  expire date: Dec 13 00:00:00 2019 GMT
*  issuer: O=Phoenix Framework; CN=Self-signed test certificate
*  SSL certificate verify result: unable to get local issuer certificate (20), continuing anyway.
* Using HTTP2, server supports multi-use
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* Using Stream ID: 1 (easy handle 0x7f80fe805200)
> GET /_rig/v1/connection/sse HTTP/2
> Host: localhost:4001
> User-Agent: curl/7.54.0
> Accept: */*
>
* Connection state changed (MAX_CONCURRENT_STREAMS updated)!
< HTTP/2 200
< access-control-allow-origin: *
< cache-control: no-cache
< content-type: text/event-stream
< date: Thu, 13 Dec 2018 22:18:15 GMT
< server: Cowboy
< x-request-id: 2lo08onm32v7hmu360000274
<
event: rig.connection.create
data: {"cloudEventsVersion":"0.1","data":{"connection_token":"g2dkAA1ub25vZGVAbm9ob3N0AAADVwAAAAAA"},"eventID":"60396982-7141-4e3b-8ebd-7890fcc981e5","eventTime":"2018-12-13T22:18:15.297959+00:00","eventType":"rig.connection.create","source":"rig"}
id: 60396982-7141-4e3b-8ebd-7890fcc981e5

event: rig.subscriptions_set
data: {"cloudEventsVersion":"0.1","data":[],"eventID":"bd088848-23d9-4082-af20-52b33a0867e9","eventTime":"2018-12-13T22:18:15.399963+00:00","eventType":"rig.subscriptions_set","source":"rig"}
id: bd088848-23d9-4082-af20-52b33a0867e9
```

# Known Issues

There are still some rough edges, but I'd rather fix them in upcoming PRs. Some known issues:

- Even though the integration tests run through, a Ranch protocol process dies along the way. I don't know anything yet about related reason and effects. The error, for reference:

```elixir
23:44:54.711 [error] Ranch protocol #PID<0.1098.0> of listener #PID<0.1075.0> (connection #PID<0.1097.0>, stream id 1) terminated
** (exit) :undef
pid=<0.1097.0>
```
- ~~Missing: test for password-protected certificate with HTTPS~~
- `:rig_common` is there but incomplete, will refactor this according to #102 next.
- The inbound gateway needs to be split up into a proxy app and an event hub app, but I don't want to make this PR any larger.
